docs/contributing/writing-instrumentation-module.md

```diff
@@ -360,3 +360,138 @@ For example:
+into a special `InstrumentationModuleClassloader`, which sees the classes from both the instrumented
+application classloader and the agent classloader.
+The instrumented classes call the advice classes residing in the `InstrumentationModuleClassloader` via
+invokedynamic bytecode instructions.
+
+### Indy modules and transition
+
+Having all instrumentation rely on native "indy" instrumentation is a tedious task and can't be
+achieved in a "big bang" step, we thus have to use intermediate steps.
+
+Instrumentation modules that are "indy native" must have their `InstrumentationModule#isIndyModule`
+implementation return `true`. Also, all the instrumentation advice methods annotated with
+`@Advice.OnMethodEnter` or `@Advice.OnMethodExit` must have the `inlined = false` explicitly set.
+
+The `otel.javaagent.experimental.indy` (default `false`) configuration option allows to opt-in for
+using "indy". When set to `true`, the `io.opentelemetry.javaagent.tooling.instrumentation.indy.AdviceTransformer`
+will transform advices automatically to make them "indy native".
+
+This configuration is automatically enabled in CI when `test indy` label is added to a pull-request
+or when the `-PtestIndy=true` parameter is added to gradle.
+
+In order to preserve compatibility with both instrumentation strategies, we have to omit the `inlined = false`
+from the advice method annotations.
+
+We have two sets of instrumentation modules:
+- "indy compatible": compatible with both "indy" and "inlined", do not override `isIndyModule`
+- "inlined only": only compatible with "inlined", `isIndyModule` returns `false`.
+
+The first step of the migration is to move all the "inlined only" to the "indy compatible" category
+by refactoring them with the limitations described below.
+
+Once everything is "indy compatible", we can evaluate changing the default value of `otel.javaagent.experimental.indy`
+to `true` and make it non-experimental.
+
+### shared classes and common classloader
+
+By default, all the advices of an instrumentation module will be loaded into isolated classloaders,
+one per instrumentation module. Some instrumentations require to use a common classloader in order
+to preserve the semantics of `static` fields and to share classes.
+
+In order to load multiple `InstrumentationModule` implementations in the same classloader, you need to
+override the `ExperimentalInstrumentationModule#getModuleGroup` to return an identical value.
+
+### advice local variables
+
+With inlined advices, declaring an advice method argument with `@Advice.Local` allows to define
+a variable that is local to the advice execution for communication between the enter and exit advices.
+
+When advices are not inlined, usage of `@Advice.Local` is not possible. It is however possible to
+return a value from the enter advice and get the value in the exit advice with a parameter annotated
+with `@Advice.Enter`, for example:
+
+```java
+@Advice.OnMethodEnter(suppress = Throwable.class)
```

Comment:
Since we're showing these examples in the context of "adapting inline advices to indy ones", I think it's worth making them fully indy compatible by adding "inline = false" to all the `@Advice.OnMethodEnter/OnMethodExit` annotations.